### ISSUE TITLE
pipeline deployment keeps re-deploying over and over again

### ISSUE THREAD

**Issue by mbalazs** [openshift/origin#11519]
I am trying to set up a build pipeline by following the tutorial on the OpenShift blog site  (https://blog.openshift.com/create-build-pipelines-openshift-3-3/), but I couldn't manage to set it up properly, not even the simple pipeline from Part 1.

When I start the pipeline, the build phase succeeds, and then at the deployment phase it keeps on deploying the application over and over again. 10 minutes and about 30 deployments later the deployment phase of the pipeline times out, and the build pipeline stops with an error. At this point, the continous re-deployment stops and I end up with a working deployment of my app.

##### Version
openshift v1.3.0
kubernetes v1.3.0+52492b4

##### Steps To Reproduce
1. Set up a pipeline build with the following config:
```yaml
apiVersion: v1
kind: BuildConfig
metadata:
  annotations:
    pipeline.alpha.openshift.io/uses: '[{"name": "myapp", "namespace": "", "kind":
      "DeploymentConfig"}]'
  creationTimestamp: null
  labels:
    name: myfirstpipeline
  name: myfirstpipeline
spec:
  output: {}
  postCommit: {}
  resources: {}
  runPolicy: Serial
  source:
    type: None
  strategy:
    jenkinsPipelineStrategy:
      jenkinsfile: |-
        node('maven') {
        stage 'build'
        openshiftBuild(buildConfig: 'myapp', showBuildLogs: 'true')
        stage 'deploy'
        openshiftDeploy(deploymentConfig: 'myapp')
        openshiftScale(deploymentConfig: 'myapp',replicaCount: '2')
        }
    type: JenkinsPipeline
  triggers: []
```
2. Deploy an S2I application ("myapp") from the web console (I used Wildfly 10 with the sample repository).
Uncheck the "Config Change" and "Image Change" triggers at the advanced options during the application creation.

3. Edit the "myapp" DC, and add the proper image spec (172.30.232.245:5000/pipeline-build/myapp:latest) on the pod template. The resulting DC looks like this:
```yaml
apiVersion: v1
kind: DeploymentConfig
metadata:
  annotations:
    openshift.io/generated-by: OpenShiftWebConsole
  creationTimestamp: null
  generation: 129
  labels:
    app: myapp
  name: myapp
spec:
  replicas: 1
  selector:
    deploymentconfig: myapp
  strategy:
    resources: {}
    rollingParams:
      intervalSeconds: 1
      maxSurge: 25%
      maxUnavailable: 25%
      timeoutSeconds: 600
      updatePeriodSeconds: 1
    type: Rolling
  template:
    metadata:
      creationTimestamp: null
      labels:
        app: myapp
        deploymentconfig: myapp
    spec:
      containers:
      - image: 172.30.232.245:5000/pipeline-build/myapp:latest
        imagePullPolicy: Always
        name: myapp
        ports:
        - containerPort: 8080
          protocol: TCP
        resources: {}
        terminationMessagePath: /dev/termination-log
      dnsPolicy: ClusterFirst
      restartPolicy: Always
      securityContext: {}
      terminationGracePeriodSeconds: 30
  test: false
  triggers: []
```
4. Start the pipeline

##### Current Result

The "Build" phase of the pipeline succeeds, but at the deployment phase the myapp DC keeps deploying over and over again until the pipeline times out. The deployments of the myapp DC always succeed. The pipeline build ends up failing after the timeout.

##### Expected Result

The deployment phase of the pipeline succeeds after a single deployment of "myapp" is completed.

**Comment by kargakis** [comment]
Can you please post the DC as is ie. with its status included?

**Comment by mbalazs** [comment]
Here is the complete DC:
```
apiVersion: v1
kind: DeploymentConfig
metadata:
  annotations:
    openshift.io/generated-by: OpenShiftWebConsole
  creationTimestamp: 2016-10-22T15:49:19Z
  generation: 129
  labels:
    app: myapp
  name: myapp
  namespace: pipeline-build
  resourceVersion: "7586093"
  selfLink: /oapi/v1/namespaces/pipeline-build/deploymentconfigs/myapp
  uid: 18344dcf-986f-11e6-9018-00163c16ea1a
spec:
  replicas: 1
  selector:
    deploymentconfig: myapp
  strategy:
    resources: {}
    rollingParams:
      intervalSeconds: 1
      maxSurge: 25%
      maxUnavailable: 25%
      timeoutSeconds: 600
      updatePeriodSeconds: 1
    type: Rolling
  template:
    metadata:
      creationTimestamp: null
      labels:
        app: myapp
        deploymentconfig: myapp
    spec:
      containers:
      - image: 172.30.232.245:5000/pipeline-build/myapp:latest
        imagePullPolicy: Always
        name: myapp
        ports:
        - containerPort: 8080
          protocol: TCP
        resources: {}
        terminationMessagePath: /dev/termination-log
      dnsPolicy: ClusterFirst
      restartPolicy: Always
      securityContext: {}
      terminationGracePeriodSeconds: 30
  test: false
  triggers:
  - type: ConfigChange
status:
  availableReplicas: 1
  details:
    causes:
    - type: ConfigChange
    message: caused by a config change
  latestVersion: 127
  observedGeneration: 129
  replicas: 1
  updatedReplicas: 1
```

And these are the last two RCs:

```
apiVersion: v1
kind: ReplicationController
metadata:
  annotations:
    openshift.io/deployer-pod.name: myapp-127-deploy
    openshift.io/deployment-config.latest-version: "127"
    openshift.io/deployment-config.name: myapp
    openshift.io/deployment.phase: Complete
    openshift.io/deployment.replicas: "1"
    openshift.io/deployment.status-reason: caused by a config change
    openshift.io/encoded-deployment-config: |
      {"kind":"DeploymentConfig","apiVersion":"v1","metadata":{"name":"myapp","namespace":"pipeline-build","selfLink":"/oapi/v1/namespaces/pipeline-build/depl
oymentconfigs/myapp","uid":"18344dcf-986f-11e6-9018-00163c16ea1a","resourceVersion":"7585277","generation":129,"creationTimestamp":"2016-10-22T15:49:19Z","lab
els":{"app":"myapp"},"annotations":{"openshift.io/generated-by":"OpenShiftWebConsole"}},"spec":{"strategy":{"type":"Rolling","rollingParams":{"updatePeriodSec
onds":1,"intervalSeconds":1,"timeoutSeconds":600,"maxUnavailable":"25%","maxSurge":"25%"},"resources":{}},"triggers":[{"type":"ConfigChange"}],"replicas":1,"t
est":false,"selector":{"deploymentconfig":"myapp"},"template":{"metadata":{"creationTimestamp":null,"labels":{"app":"myapp","deploymentconfig":"myapp"}},"spec
":{"containers":[{"name":"myapp","image":"172.30.232.245:5000/pipeline-build/myapp:latest","ports":[{"containerPort":8080,"protocol":"TCP"}],"resources":{},"t
erminationMessagePath":"/dev/termination-log","imagePullPolicy":"Always"}],"restartPolicy":"Always","terminationGracePeriodSeconds":30,"dnsPolicy":"ClusterFir
st","securityContext":{}}}},"status":{"latestVersion":127,"observedGeneration":128,"replicas":1,"updatedReplicas":1,"availableReplicas":1,"details":{"message"
:"caused by a config change","causes":[{"type":"ConfigChange"}]}}}
    openshift.io/jenkins-build-uri: https://jenkins-pipeline-build.apps.podspace.hu/job/myfirstpipeline/9/
  creationTimestamp: 2016-10-24T09:01:15Z
  generation: 2
  labels:
    app: myapp
    openshift.io/deployment-config.name: myapp
  name: myapp-127
  namespace: pipeline-build
  resourceVersion: "7586102"
  selfLink: /api/v1/namespaces/pipeline-build/replicationcontrollers/myapp-127
  uid: 6b457b6c-99c8-11e6-9018-00163c16ea1a
spec:
  replicas: 1
  selector:
    deployment: myapp-127
    deploymentconfig: myapp
  template:
    metadata:
      annotations:
        openshift.io/deployment-config.latest-version: "127"
        openshift.io/deployment-config.name: myapp
        openshift.io/deployment.name: myapp-127
      creationTimestamp: null
      labels:
        app: myapp
        deployment: myapp-127
        deploymentconfig: myapp
    spec:
      containers:
      - image: 172.30.232.245:5000/pipeline-build/myapp:latest
        imagePullPolicy: Always
        name: myapp
        ports:
        - containerPort: 8080
          protocol: TCP
        resources: {}
        terminationMessagePath: /dev/termination-log
      dnsPolicy: ClusterFirst
      restartPolicy: Always
      securityContext: {}
      terminationGracePeriodSeconds: 30
status:
  fullyLabeledReplicas: 1
  observedGeneration: 2
  replicas: 1
```
and the one before that:
```
apiVersion: v1
kind: ReplicationController
metadata:
  annotations:
    kubectl.kubernetes.io/original-replicas: "1"
    openshift.io/deployer-pod.name: myapp-126-deploy
    openshift.io/deployment-config.latest-version: "126"
    openshift.io/deployment-config.name: myapp
    openshift.io/deployment.phase: Complete
    openshift.io/deployment.replicas: "0"
    openshift.io/deployment.status-reason: caused by a config change
    openshift.io/encoded-deployment-config: |
      {"kind":"DeploymentConfig","apiVersion":"v1","metadata":{"name":"myapp","namespace":"pipeline-build","selfLink":"/oapi/v1/namespaces/pipeline-build/depl
oymentconfigs/myapp","uid":"18344dcf-986f-11e6-9018-00163c16ea1a","resourceVersion":"7585124","generation":128,"creationTimestamp":"2016-10-22T15:49:19Z","lab
els":{"app":"myapp"},"annotations":{"openshift.io/generated-by":"OpenShiftWebConsole"}},"spec":{"strategy":{"type":"Rolling","rollingParams":{"updatePeriodSec
onds":1,"intervalSeconds":1,"timeoutSeconds":600,"maxUnavailable":"25%","maxSurge":"25%"},"resources":{}},"triggers":[{"type":"ConfigChange"}],"replicas":1,"t
est":false,"selector":{"deploymentconfig":"myapp"},"template":{"metadata":{"creationTimestamp":null,"labels":{"app":"myapp","deploymentconfig":"myapp"}},"spec
":{"containers":[{"name":"myapp","image":"172.30.232.245:5000/pipeline-build/myapp:latest","ports":[{"containerPort":8080,"protocol":"TCP"}],"resources":{},"t
erminationMessagePath":"/dev/termination-log","imagePullPolicy":"Always"}],"restartPolicy":"Always","terminationGracePeriodSeconds":30,"dnsPolicy":"ClusterFir
st","securityContext":{}}}},"status":{"latestVersion":126,"observedGeneration":127,"replicas":1,"updatedReplicas":1,"availableReplicas":1,"details":{"message"
:"caused by a config change","causes":[{"type":"ConfigChange"}]}}}
    openshift.io/jenkins-build-uri: https://jenkins-pipeline-build.apps.podspace.hu/job/myfirstpipeline/9/
  creationTimestamp: 2016-10-24T09:00:45Z
  generation: 3
  labels:
    app: myapp
    openshift.io/deployment-config.name: myapp
  name: myapp-126
  namespace: pipeline-build
  resourceVersion: "7586104"
  selfLink: /api/v1/namespaces/pipeline-build/replicationcontrollers/myapp-126
  uid: 59342852-99c8-11e6-9018-00163c16ea1a
spec:
  replicas: 0
  selector:
    deployment: myapp-126
    deploymentconfig: myapp
  template:
    metadata:
      annotations:
        openshift.io/deployment-config.latest-version: "126"
        openshift.io/deployment-config.name: myapp
        openshift.io/deployment.name: myapp-126
      creationTimestamp: null
      labels:
        app: myapp
        deployment: myapp-126
        deploymentconfig: myapp
    spec:
      containers:
      - image: 172.30.232.245:5000/pipeline-build/myapp:latest
        imagePullPolicy: Always
        name: myapp
        ports:
        - containerPort: 8080
          protocol: TCP
        resources: {}
        terminationMessagePath: /dev/termination-log
      dnsPolicy: ClusterFirst
      restartPolicy: Always
      securityContext: {}
      terminationGracePeriodSeconds: 30
status:
  observedGeneration: 3
  replicas: 0
```

**Comment by kargakis** [comment]
So the only change between versions of the DC seems to be the latestVersion bump. Templates between deployment configs are equal so this is not caused by the trigger controller who is responsible for triggering a new rollout on a pod template change (config change). @gabemontero are you manually updating DCs with the Jenkins plugin?

**Comment by kargakis** [comment]
I assume that's the second before last step triggered in the pipeline.
```
node('maven') {
        stage 'build'
        openshiftBuild(buildConfig: 'myapp', showBuildLogs: 'true')
        stage 'deploy'
        openshiftDeploy(deploymentConfig: 'myapp')
        openshiftScale(deploymentConfig: 'myapp',replicaCount: '2')
}
```
So it seems that every time the deployment is completed, Jenkins retriggers it.

**Comment by gabemontero** [comment]
I have not been able to reproduce trying various levels, but I have a couple of points (both a possible existing fixes and debug if not fixed).

1) to @kargakis 's question, yes, based on REST flow analysis of `oc deploy <dc name> --latest` when the deploy step was first constructed last year, the plugin deploy step would manually update the DC's latest version.  However, starting with v1.0.23 of the openshift-pipeline plugin, the plugin first attempts to use the instantiate endpoint if it exists.

@mbalazs - when trying this scenario, log into your Jenkins UI and navigate to Manage Jenkins -> Manage Plugins and click the Installed tab, and see what version of OpenShift Pipeline you have.  If it is prior to v1.0.23, you have a couple of upgrade options
- if using the openshift jenkins centos image, simply pull down the latest from docker hub
- if using the openshift jenkins rhel image, grabbing a more recent rhel image will depend on what you have access to ... updating the plugin directly from the plugin manager might be the easiest option.  Here is the link to the jenkins update center for the plugin:  https://updates.jenkins-ci.org/download/plugins/openshift-pipeline/

2) If the problem is still happening with a later version of the plugin, I'll try a bit more to reproduce on my own, but if I have no luck, I'll need a repro from you with debug in place.  To turn on debug, change your last two steps to:
```
openshiftDeploy(deploymentConfig: 'myapp', verbose: 'true')
openshiftScale(deplomentConfig: 'myapp', replicaCount: '2', verbose: 'true')
```
and provide me the output

thanks

**Comment by kargakis** [comment]
> 1) to @kargakis 's question, yes, based on REST flow analysis of oc deploy <dc name> --latest when the deploy step was first constructed last year, the plugin deploy step would manually update the DC's latest version. However, starting with v1.0.23 of the openshift-pipeline plugin, the plugin first attempts to use the instantiate endpoint if it exists.

Thanks. This means that in the case of @mbalazs the Jenkins plugin will just update status.latestVersion directly in the DC since the running version of OpenShift is 1.3 and instantiate will be included in 1.4.

**Comment by gabemontero** [comment]
Breaking update ... I think I've managed to reproduced.  Investigating.

@bparees FYI - I believe this is independent of the recent features dropped this sprint.

**Comment by gabemontero** [comment]
OK I have validated a fix.  There is a timing issue wrt obtaining the RC after bumping the DC, where if the RC is not ready in time, we can get in the loop described.  Timing window seemed more prevalent going against 1.3 vs. 1.4, but the necessary fix is in the plugin.

As such, the plugin is in a separate repo, so I can't automatically close out this issue via the PR there.  So closing this issue out now in favor of  https://github.com/openshift/jenkins-plugin/pull/98

The fix will be in v1.0.27 and later of the openshift-pipeline plugin.
